docs/spec_draft.md

```diff
@@ -335,6 +336,127 @@ For boolean tensors, computes the logical operation.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.batch_norm_grad
+
+### Semantics
+
+Computes gradients of `operand`, `scale` and offset
+(refer [batch_norm_training](#stablehlobatch_norm_training)), using `mean`,
+`variance` and `grad_output`, the gradient of output from
+[batch_norm_training](#stablehlobatch_norm_training), and produces
+`grad_operand`, `grad_scale` and `grad_offset` tensors. More formally, this
+operation can be expressed as a decomposition to existing StableHLO operations
+using Python-like syntax as follows:
+
+```python
+def reduce_sum(operand, feature_index):
+  (sum,) = reduce(
+      inputs=[operand],
+      init_values=[0.0],
+      dimensions=[i for i in range(rank(operand)) if i != feature_index],
+      body=lambda x, y: add(x, y))
+  return sum
+
+def compute_mean(operand, feature_index):
+  sum = reduce_sum(operand, feature_index)
+  divisor = constant(num_elements(operand) / dim(operand, feature_index))
+  divisor_bcast = broadcast_in_dim(divisor, [], shape(sum))
+  return divide(sum, divisor_bcast)
+
+def batch_norm_grad(operand, scale, mean, variance, grad_output, epsilon, feature_index):
+  # Broadcast inputs to shape(operand)
+  scale_bcast = broadcast_in_dim(scale, [feature_index], shape(operand))
+  mean_bcast = broadcast_in_dim(mean, [feature_index], shape(operand))
+  variance_bcast = broadcast_in_dim(variance, [feature_index], shape(operand))
+  epsilon_bcast = broadcast_in_dim(constant(epsilon), [], shape(operand))
+
+  centered_operand = subtract(operand, mean_bcast)
+  stddev = sqrt(add(variance_bcast, epsilon_bcast))
+  normalized_operand = divide(centered_operand, stddev)
+
+  intermediate_expr = compute_mean(
+    multiply(grad_output, divide(normalized_operand, stddev)), feature_index)
+  intermediate_expr_bcast = broadcast_in_dim(
+    intermediate_expr, [feature_index], shape(operand))
+  mean_grad_output = compute_mean(grad_output, feature_index)
+  mean_grad_output_bcast = broadcast_in_dim(
+    mean_grad_output, [feature_index], shape(operand))
+
+  grad_operand = multiply(
+    divide(scale_bcast, stddev),
+    subtract(
+      subtract(grad_output, mean_grad_output_bcast),
+      multiply(intermediate_expr_bcast, centered_operand)
+    )
+  )
+  grad_scale = reduce_sum(
+    multiply(grad_output, normalized_operand), feature_index)
+  grad_offset = reduce_sum(grad_output, feature_index)
+  return grad_operand, grad_scale, grad_offset
+```
```

Comment:
Proposed reformulation:

```Python
def compute_sum(operand, feature_index):
  (sum,) = reduce(
      inputs=[operand],
      init_values=[0.0],
      dimensions=[i for i in range(rank(operand)) if i != feature_index],
      body=lambda x, y: add(x, y))
  return sum

def compute_mean(operand, feature_index):
  sum = compute_sum(operand, feature_index)
  divisor = constant(num_elements(operand) / dim(operand, feature_index))
  divisor_bcast = broadcast_in_dim(divisor, [], shape(sum))
  return divide(sum, divisor_bcast)

def batch_norm_grad(operand, scale, mean, variance, grad_output, epsilon, feature_index):
  # Broadcast inputs to shape(operand)
  scale_bcast = broadcast_in_dim(scale, [feature_index], shape(operand))
  mean_bcast = broadcast_in_dim(mean, [feature_index], shape(operand))
  variance_bcast = broadcast_in_dim(variance, [feature_index], shape(operand))
  epsilon_bcast = broadcast_in_dim(constant(epsilon), [], shape(operand))

  # Perform normalization using the provided `mean` and `variance`
  # Intermediate values will be useful for computing gradients
  centered_operand = subtract(operand, mean_bcast)
  stddev = sqrt(add(variance_bcast, epsilon_bcast))
  normalized_operand = divide(centered_operand, stddev)

  # Use the implementation from batchnorm_expander.cc in XLA
  # Temporary variables have exactly the same names as in the C++ code
  elements_per_feature = constant(dim(operand, feature_index))
  i1 = multiply(
    grad_output, 
    broadcast_in_dim(elements_per_feature, [], shape(operand)))
  i2 = broadcast_in_dim(
    compute_sum(grad_output, feature_index), 
    [feature_index], shape(operand))
  i3 = broadcast_in_dim(
    compute_sum(multiply(grad_output, centered_operand)),
    [feature_index], shape(operand))
  i4 = multiply(i3, centered_operand)
  i5 = divide(i4, add(variance_bcast, epsilon_bcast))
  grad_operand = multiply(
    divide(scale_bcast, stddev),
    subtract(subtract(i1, i2), i5))
  )
  grad_scale = compute_sum(
    multiply(grad_output, normalized_operand), feature_index)
  grad_offset = compute_sum(grad_output, feature_index)
  return grad_operand, grad_scale, grad_offset
```

Feedback:
  1. Renamed `reduce_sum` to `compute_sum` to be consistent with other `compute_foo` functions in BatchNorm specs.
  2. Used the implementation for `grad_operand` from batchnorm_expander.cc in XLA, keeping exactly the same temporary variable names: i1, i2, i3, i4 and i5. As you wrote in a nearby comment, these variables don't have good names, so let's align them with something existing for the ease of reviews.

Reply:
```elements_per_feature = constant(
    divide(size(operand), dim(operand, feature_index)))```

Reply:
```python
grad_operand = multiply(
    divide(divide(scale_bcast, stddev), elements_per_feature),
    subtract(subtract(i1, i2), i5))
```

Reply:
I had thoroughly checked the previous formulation with `batchnorm_expander` for equivalence.

Reply:
```python
def batch_norm_grad(operand, scale, mean, variance, grad_output, epsilon, feature_index):
  # Broadcast inputs to shape(operand)
  scale_bcast = broadcast_in_dim(scale, [feature_index], shape(operand))
  mean_bcast = broadcast_in_dim(mean, [feature_index], shape(operand))
  variance_bcast = broadcast_in_dim(variance, [feature_index], shape(operand))
  epsilon_bcast = broadcast_in_dim(constant(epsilon), [], shape(operand))
  centered_operand = subtract(operand, mean_bcast)
  stddev = sqrt(add(variance_bcast, epsilon_bcast))
  normalized_operand = divide(centered_operand, stddev)
  intermediate_expr = compute_mean(
    multiply(grad_output, divide(normalized_operand, stddev)), feature_index)
  intermediate_expr_bcast = broadcast_in_dim(
    intermediate_expr, [feature_index], shape(operand))
  mean_grad_output = compute_mean(grad_output, feature_index)
  mean_grad_output_bcast = broadcast_in_dim(
    mean_grad_output, [feature_index], shape(operand))
  grad_operand = multiply(
    divide(scale_bcast, stddev),
    subtract(
      subtract(grad_output, mean_grad_output_bcast),
      multiply(intermediate_expr_bcast, centered_operand)
    )
  )
  grad_scale = compute_sum(
    multiply(grad_output, normalized_operand), feature_index)
  grad_offset = compute_sum(grad_output, feature_index)
  return grad_operand, grad_scale, grad_offset
```

Reply:
This is exactly same as outlined in [XLA semantics](https://www.tensorflow.org/xla/operation_semantics#batchnormgrad)

Reply:
Thank you for doing the hard work going through the two implementations of BatchNormGrad, making sure they are equivalent and then writing things down!

As we discussed offline, it would be preferable to choose the expander-based version because it's easier to review (since it's running in production, we can be certain that it's correct, while with the version from the Operation Semantics page, figuring that out requires non-trivial work), but I really appreciate the attention to detail.

Agreed with both of your proposed changes! This is something that I missed.